contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,368 @@ const (
+  type: pd-standard
+  replicas: 2
+allowedTopologies:
+  requiredToplogies:
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-a
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-b
+```
+
+#### Multi Label Example
+This example restricts the volume's topology to nodes that
+have the following labels:
+
+* "zone: us-central1-a" and "rack: rack1" or,
+* "zone: us-central1-b" and "rack: rack1" or,
+* "zone: us-central1-b" and "rack: rack2"
+
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: something-fancy
+provisioner: rack-based-provisioner
+parameters:
+allowedTopologies:
+  requiredTopologies:
+  - matchNodeLabels:
+      zone: us-central1-a
+      rack: rack1
+  - matchNodeLabels:
+      zone: us-central1-b
+      rack: rack1
+  - matchNodeLabels:
+      zone: us-central1-b
+      rack: rack2
+```
+
+### Storage Provider Capacity Reporting
+Some volume types have capacity limitations per topology domain. For example,
+local volumes have limited capacity per node. Provisioners for these volume
+types can report these capacity limitations through a new StorageClass status
+field.
+
+```
+type StorageClass struct {
+    ...
+
+    Status *StorageClassStatus
+}
+
+type StorageClassStatus struct {
+    Capacity *StorageClassCapacity
+}
+
+type StorageClassCapacity struct {
+    // The node label key that the provisioner operates on
+    TopologyKey string
+    // Capacities specifies the total capacity available
+    // (free and used) for provisioning at each value of the
+    // TopologyKey
+    Capacities map[string]resource.Quantity
+}
+```
+
+If no StorageClassCapacity is specified, then there are no capacity
+limitations.
+
+Capacity reporting is limited to a single topology key. `TopologyKey`
+specifies the label key that represents that domain. Capacities is
+a map with key equal to the label value, and value is the capacity for that
+specific domain. Capacity is the total capacity available, including
+capacity that has already been provisioned. A missing map entry for a topology
+domain will be treated as zero capacity.
+
+A scheduler predicate will use the reported capacity to determine if dynamic
+provisioning is possible for a given node, taking into account already
+provisioned capacity.
+
+Downsides of this approach include:
+* Multiple topology keys is not supported for capacity reporting. We have not
+  seen such use cases so far.
+* There is no common mechanism to report capacity. Each provisioner has to
+  update their own fields in the StorageClass if necessary. CSI could potentially
+  address this.
+* There is a gap during cluster initialization where the StorageClass object
+  has been created, and the provisioner has not reported capacity yet. In this
+  scenario, the provisioner should fail provisioning until it has reported its
+  capacity and not process any PVCs that have not been updated with the chosen
+  topologyKey value.
+
```

Comment:
Another downsides:

- When the storage is shared among several Kubernetes clusters *or* with something outside Kubernetes, Kubernetes cannot reliably compute amount of used storage. It can compute only the storage the particular cluster consumes (i.e. PVs), but it won't see PVs of other clusters and/or users outside of Kubernetes. It is recommended that Kubernetes gets dedicated storage or storage quota so it can reliably compute amount of used/remaining storage just by counting PVs.

- The storage may be shared by several StorageClasses. Typically, Gluster allows admins to choose nr. of replicas of PVs in StorageClass. There may be two StorageClasses that use the same Gluster servers, however each of them creates PVs with different nr. of replicas. Kubernetes treats StorageClasses as independent on each other, however in this case PVs in one StorageClass must be counted to used space in a different StorageClass.

- Capacity does not report granularity of the storage. E.g. there may be 1 GiB of available space on Gluster in a particular datacenter and scheduler can schedule one pod with two 0.5 GiB PVCs there. Gluster allocates storage in 1GiB chunks, so it can't accommodate both PVCs and the pod becomes unschedulable until someone frees up some space in the datacenter.
  - This is IMO rare corner case when a storage is almost full.

Reply:
> I propose to instead probe the storage provider if needed: "Could you create a volume of size X and type Y?" This could be a "dry run" variant of the volume create api.

Scheduler is under watch for its performance. Currently, it does not call any external API. Calling number of probe/dry-run methods would slow everything down. We need a quick and mostly reliable check where to start provisioning first, hoping it succeeds in vast majority of cases.

Reply:
Another problem with doing a call out is that the "provisioning check" and "actual provisioning" steps are not synchronous.  So the following two scenarios would not be accurate:

* Pod requests multiple PVCs.  You would need to query the storage system in a way so that the 1st query would temporarily subtract from the total when doing the 2nd query.
* Scheduler does the actual binding operations asynchronously, so you could have a query for the next Pod before the provisioning for the 1st Pod is triggered.  This is similar to the 1st scenario, just for different Pods.

That's why the scheduler today doesn't do callouts to get the Node resources, and it tracks the used capacity internally.  So if we want to do something similar for storage, then I think we should follow this similar model of storing total capacity in an object.  I agree that there are many internal details that influence what is actually provisionable, so if we had such an API, it would be up to the provisioner to report its capacity wrt to the user's request.  However, the limitation of the current capacity API is that the storage pool is contiguous, which doesn't sound to be the case with gluster (discrete disks).